File: src/platform/gtk/toolbar.c
/* Claro Graphics - an abstraction layer for native UI libraries
 * 
 * $Id$
 * 
 * The contents of this file are subject to the Mozilla Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 * 
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 * 
 * See the LICENSE file for more details.
 */


#include "graphics.h"
#include <assert.h>
#include "platform.h"

/* Toolbar */

void cgraphics_toolbar_item_pushed_handler( GtkToolItem *gitem, list_item_t *i )
{
	event_send( OBJECT(i), "pushed", "" );
}

void cgraphics_toolbar_widget_create( widget_t *widget )
{
	widget_t *parent;
	gtk_window_native *nd;
	widget->native = gtk_toolbar_new( );
	widget->ndata = gtk_tooltips_new( );
	
	if ( widget->flags & cToolbarShowBoth )
		gtk_toolbar_set_style( GTK_TOOLBAR(widget->native), GTK_TOOLBAR_BOTH );
	else if ( widget->flags & cToolbarShowText )
		gtk_toolbar_set_style( GTK_TOOLBAR(widget->native), GTK_TOOLBAR_TEXT );
	else
		gtk_toolbar_set_style( GTK_TOOLBAR(widget->native), GTK_TOOLBAR_ICONS );
	
	parent = (widget_t *)widget->object.parent;
	
	if ( parent == 0 )
		return;
	
	if ( strcasecmp( parent->object.type, "claro.graphics.widgets.window" ) )
		return;
	
	nd = parent->ndata;
	
	gtk_box_pack_end( GTK_BOX(nd->vboxtop), GTK_WIDGET(widget->native), FALSE, FALSE, 0 );
}

void cgraphics_toolbar_new_icon( widget_t *widget, list_item_t *item )
{
	image_t *img = item->data[0];
	char *title = item->data[1];
	char *tooltip = item->data[2];
	GtkToolItem *ti;
	
	if ( img != 0 || title != 0 )
	{
		GtkWidget *icon = NULL;
		
		if ( img != 0 )
		{
			icon = gtk_image_new_from_pixbuf( img->native );
			gtk_widget_show( icon );
		}
		
		ti = gtk_tool_button_new( icon, title );
		
		if ( tooltip != 0 )
		{
			int a = 0;
			char tmpt[strlen(tooltip)+1];
			
			// kill mnemonics..
			while ( *tooltip != 0 )
			{
				if ( *tooltip != '&' )
				{
					tmpt[a] = *tooltip;
					a++;
				}
				
				tooltip++;
			}
			tmpt[a] = 0;

			gtk_tool_item_set_tooltip( ti, widget->ndata, tmpt, "" );
		}
	}
	else
	{
		ti = gtk_separator_tool_item_new();
	}
	
	item->native = ti;
	
	gtk_widget_show( GTK_WIDGET(ti) );
	
	gtk_toolbar_insert( widget->native, ti, item->row );
	
	if ( img != 0 || title != 0 )
		g_signal_connect( G_OBJECT(ti), "clicked", G_CALLBACK(cgraphics_toolbar_item_pushed_handler), item );
}

void cgraphics_toolbar_remove_icon( widget_t *widget, list_item_t *item )
{
	
}

void cgraphics_toolbar_set_menu( widget_t *widget, list_item_t *item )
{
	image_t *img = item->data[0];
	char *title = item->data[1];
	char *tooltip = item->data[2];
	GtkToolItem *oti = item->native;
	GtkToolItem *ti;
	
	if ( img == 0 && title == 0 )
		return;
	
	if ( item->menu != NULL )
	{
		GtkWidget *icon = NULL;
		
		if ( img != 0 )
		{
			icon = gtk_image_new_from_pixbuf( img->native );
			gtk_widget_show( icon );
		}
		
		gtk_widget_destroy( GTK_WIDGET(oti) );
		
		ti = gtk_menu_tool_button_new( icon, title );
		gtk_menu_tool_button_set_menu( GTK_MENU_TOOL_BUTTON(ti), WIDGET(item->menu)->native );
		
		item->native = ti;
		
		if ( tooltip != 0 )
		{
			int a = 0;
			char tmpt[strlen(tooltip)+1];
			
			// kill mnemonics..
			while ( *tooltip != 0 )
			{
				if ( *tooltip != '&' )
				{
					tmpt[a] = *tooltip;
					a++;
				}
				
				tooltip++;
			}
			tmpt[a] = 0;
			
			gtk_tool_item_set_tooltip( ti, widget->ndata, tmpt, "" );
		}
	}
	else
		return;
	
	gtk_widget_show( GTK_WIDGET(ti) );
	
	gtk_toolbar_insert( widget->native, ti, item->row );
	
	if ( img != 0 || title != 0 )
		g_signal_connect( G_OBJECT(ti), "clicked", G_CALLBACK(cgraphics_toolbar_item_pushed_handler), item );
}


/*
void cgraphics_toolbar_new_popup_item( widget_t *widget, list_item_t *item )
{
	
}

void cgraphics_toolbar_remove_popup_item( widget_t *widget, list_item_t *item )
{
	
}*/

File: src/platform/cocoa/toolbar.m
/* Claro Graphics - an abstraction layer for native UI libraries
 * 
 * $Id$
 * 
 * The contents of this file are subject to the Mozilla Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 * 
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 * 
 * See the LICENSE file for more details.
 */


#define _PLATFORM_INC
#include "graphics.h"
#include "platform/macosx_cocoa.h"

/* Toolbar */

list_item_t *cgraphics_find_list_item_by_id( list_t *list, NSString *id );

/* ClaroToolBar (subclassed from NSToolbar) */
@interface ClaroToolBar : NSToolbar
{
	widget_t *cw;
}

/* internal init function */
- (void)setClaroWidget:(widget_t *)widget;

/* notification responders */
- (void)claroResize:(NSNotification *)aNotification;
- (void)claroClose:(NSNotification *)aNotification;
	
- (NSToolbarItem *)toolbar:(NSToolbar *)toolbar
	itemForItemIdentifier:(NSString *)itemIdentifier
	willBeInsertedIntoToolbar:(BOOL)flag;

- (NSArray *)toolbarAllowedItemIdentifiers:(NSToolbar*)toolbar;
@end

@implementation ClaroToolBar

- (void)claroResize:(NSNotification *)aNotification
{
	
}

- (void)claroClose:(NSNotification *)aNotification
{
	widget_destroy( cw );
}

- (void)setClaroWidget:(widget_t *)widget
{
	cw = widget;
	
	[self setDelegate:self];
	[self setAllowsUserCustomization:NO];
	[self setAutosavesConfiguration: NO]; 
	
	if ( cw->flags & (cToolbarShowImages|cToolbarShowText) )
		[self setDisplayMode: NSToolbarDisplayModeIconAndLabel];
	else if ( cw->flags & cToolbarShowImages )
		[self setDisplayMode: NSToolbarDisplayModeIconOnly];
	else
		[self setDisplayMode: NSToolbarDisplayModeLabelOnly];
	
	[[NSNotificationCenter defaultCenter] addObserver:self
		selector:@selector(claroClose:) name:NSWindowWillCloseNotification
		object:self];
	
	[[NSNotificationCenter defaultCenter] addObserver:self
		selector:@selector(claroResize:) name:NSViewFrameDidChangeNotification
		object:self];
}

- (NSToolbarItem *)toolbar:(NSToolbar *)toolbar
	itemForItemIdentifier:(NSString *)itemIdentifier
	willBeInsertedIntoToolbar:(BOOL)flag
{
	NSToolbarItem *item = [[NSToolbarItem alloc] initWithItemIdentifier:itemIdentifier];
	list_widget_t *lw = (list_widget_t *)cw;
	list_item_t *litem = cgraphics_find_list_item_by_id( &lw->items, itemIdentifier );
	
	if ( litem != NULL )
	{
		image_t *img = litem->data[0];
		const char *txt = litem->data[1];
		const char *ttip = litem->data[2];
		
		if ( txt != 0 )
			[item setLabel:[[NSString alloc] initWithCString:txt encoding:NSUTF8StringEncoding]];
		
		if ( ttip != 0 )
			[item setToolTip:[[NSString alloc] initWithCString:ttip encoding:NSUTF8StringEncoding]];
		
		if ( img != 0 )
			[item setImage:img->native];
	}
	
	return [item autorelease];
}

- (NSArray *)toolbarAllowedItemIdentifiers:(NSToolbar*)toolbar
{
	return [NSArray arrayWithObjects:NSToolbarSeparatorItemIdentifier, nil];
}

@end

list_item_t *cgraphics_find_list_item_by_id( list_t *list, NSString *id )
{
	node_t *n;
	list_item_t *item;
	
	LIST_FOREACH( n, list->head )
	{
		item = (list_item_t *)n->data;
		
		if ( [id isEqualToString:(NSString *)item->nativeid] )
			return item;
		
		item = cgraphics_find_list_item_by_id( &item->children, id );
		
		if ( item != NULL )
			return item;
	}
	
	return NULL;
}

void cgraphics_toolbar_widget_create( widget_t *widget )
{
	ClaroToolBar *tb;
	NSView *parent = cgraphics_get_native_parent( widget );
	NSWindow *win = [parent window];
	
	tb = [[ClaroToolBar alloc]
		initWithIdentifier:@"claroToolbar"
		];
	
	[tb setClaroWidget:widget];
	[win setToolbar:[tb autorelease]];
	//[parent addSubview: tb];
	
	widget->native = (NSControl *)tb;
}

void cgraphics_toolbar_new_icon( widget_t *widget, list_item_t *item )
{
	ClaroToolBar *tb = widget->native;
	//NSToolbarItem *titem;
	NSString *str;
	image_t *img = item->data[0];
	const char *txt = item->data[1];
	//const char *ttip = item->data[2];
	char tmp[512];
	
	if ( img == 0 && txt == 0 )
	{
		/* regardless of style, this is a separator :) */
		
		//item = [[[NSToolbarItem alloc] initWithItemIdentifier:NSToolbarSeparatorItemIdentifier] autorelease];
		[tb insertItemWithItemIdentifier:NSToolbarSeparatorItemIdentifier atIndex:item->row ];
		return;
	}
	
	sprintf( tmp, "clarotoolbaritem%p", item );

	str = [[NSString alloc] initWithCString:tmp encoding:NSUTF8StringEncoding];
	item->nativeid = (int)str;
	
	[tb insertItemWithItemIdentifier:str atIndex:item->row ];
}

void cgraphics_toolbar_remove_icon( widget_t *widget, list_item_t *item )
{
	
}

void cgraphics_toolbar_set_menu( widget_t *widget, list_item_t *item )
{
	
}
